system/lib/libunwind/src/Unwind-wasm.cpp

```diff
@@ -0,0 +1,110 @@
+#include "config.h"
+#include "unwind.h"
+#include <stdbool.h>
+
+#ifdef __USING_WASM_EXCEPTIONS__
+
+extern "C" {
+
+_Unwind_Reason_Code __gxx_personality_v0(int version, _Unwind_Action actions,
+                                         uint64_t exceptionClass,
+                                         _Unwind_Exception *unwind_exception,
+                                         _Unwind_Context *context);
+
+struct _Unwind_LandingPadContext {
+  // Input information to personality function
+  uintptr_t lpad_index; // landing pad index
+  uintptr_t lsda;       // LSDA address
+
+  // Output information computed by personality function
+  uintptr_t selector; // selector value
+};
+
+// Communication channel between WebAssembly code and personality function
+struct _Unwind_LandingPadContext __wasm_lpad_context;
+
+/// Calls to this function is in landing pads in compiler-generated user code.
+/// In other EH schemes, stack unwinding is done by libunwind library, which
+/// calls the personality function for each each frame it lands. On the other
+/// hand, WebAssembly stack unwinding process is performed by a VM, and the
+/// personality function cannot be called from there. So the compiler inserts
+/// a call to this function in landing pads in the user code, which in turn
+/// calls the personality function.
+_Unwind_Reason_Code _Unwind_CallPersonality(void *exception_ptr) {
+  struct _Unwind_Exception *exception_object =
+      (struct _Unwind_Exception *)exception_ptr;
+  _LIBUNWIND_TRACE_API("_Unwind_CallPersonality(exception_object=%p)",
+                       (void *)exception_object);
+
+  // Call personality function. Wasm does not have two-phase unwinding, so we
+  // only do the cleanup phase.
+  _Unwind_Reason_Code ret = __gxx_personality_v0(
+      1, _UA_CLEANUP_PHASE, exception_object->exception_class, exception_object,
+      (struct _Unwind_Context *)&__wasm_lpad_context);
+  return ret;
+}
+
+/// Called by __cxa_throw.
+_LIBUNWIND_EXPORT _Unwind_Reason_Code
+_Unwind_RaiseException(_Unwind_Exception *exception_object) {
+  _LIBUNWIND_TRACE_API("_Unwind_RaiseException(exception_object=%p)",
+                       (void *)exception_object);
+  __builtin_wasm_throw(0, exception_object);
+}
+
+/// Called by __cxa_end_catch.
+_LIBUNWIND_EXPORT void
+_Unwind_DeleteException(_Unwind_Exception *exception_object) {
+  _LIBUNWIND_TRACE_API("_Unwind_DeleteException(ex_obj=%p)",
+                       (void *)(exception_object));
+  if (exception_object->exception_cleanup != NULL)
+    (*exception_object->exception_cleanup)(_URC_FOREIGN_EXCEPTION_CAUGHT,
```

Comment:
This code was copied from other targets (the implementation of this function looks the same for all other targets). But I haven't tested this function myself, because we don't really have a test case for foreign exceptions now. I added this to avoid linking failure. The reason this is about foreign exception is, the only place this function is called is the part in `__cxa_end_catch` that deals with foreign exceptions.